distributed/shuffle/shuffle_extension.py

```diff
@@ -0,0 +1,348 @@
+from __future__ import annotations
+
+import asyncio
+import math
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Callable, NewType
+
+import pandas as pd
+
+from distributed.protocol import to_serialize
+from distributed.utils import sync
+
+if TYPE_CHECKING:
+    from distributed.worker import Worker
+
+ShuffleId = NewType("ShuffleId", str)
+
+
+# NOTE: we use these dataclasses primarily for type-checking benefits.
+# They take the place of positional arguments to `shuffle_init`,
+# which the type-checker can't validate when it's called as an RPC.
+
+
+@dataclass(frozen=True, eq=False)
+class NewShuffleMetadata:
+    "Metadata to create a shuffle"
+    id: ShuffleId
+    empty: pd.DataFrame
+    column: str
+    npartitions: int
+
+
+@dataclass(frozen=True, eq=False)
+class ShuffleMetadata(NewShuffleMetadata):
+    """
+    Metadata every worker needs to share about a shuffle.
+
+    A `ShuffleMetadata` is created with a task and sent to all workers
+    over the `ShuffleWorkerExtension.shuffle_init` RPC.
+    """
+
+    workers: list[str]
+
+    def worker_for(self, output_partition: int) -> str:
+        "Get the address of the worker which should hold this output partition number"
+        assert output_partition >= 0, f"Negative output partition: {output_partition}"
+        if output_partition >= self.npartitions:
+            raise IndexError(
+                f"Output partition {output_partition} does not exist in a shuffle producing {self.npartitions} partitions"
+            )
+        i = len(self.workers) * output_partition // self.npartitions
+        return self.workers[i]
+
+    def partition_range(self, worker: str) -> tuple[int, int]:
+        "Get the output partition numbers (inclusive) that a worker will hold"
+        i = self.workers.index(worker)
+        first = math.ceil(self.npartitions * i / len(self.workers))
+        last = math.ceil(self.npartitions * (i + 1) / len(self.workers)) - 1
+        return first, last
```

Comment:
I'm not sure what you mean? Do you think this test of it is missing something?
https://github.com/dask/distributed/blob/0e403ec9364b720c8469ddda74199004fb601cbe/distributed/shuffle/tests/test_shuffle_extension.py#L52-L61

Reply:
I'm mostly wondering why a worker has a continuous range of IDs. I guess this connects to how we calculate the hash and partition ID. Can you put that information. Is it documented somewhere how all of these things connect and which ID is which and how they are calcualted, etc.

Reply:
> I'm mostly wondering why a worker has a continuous range of IDs.

That's an implementation detail of the binning arithmetic (`//` instead of `%`). I'll make this method private. It should not be required, nor relied on as part of the shuffle interface, that workers receive consecutive output partitions (though it is good for scheduling). This method just made it more readable and intuitive to implement `npartitions_for`, which _is_ needed.